src/openssl.c

```diff
@@ -1679,6 +2571,244 @@ _libssh2_ecdh_gen_k(_libssh2_bn **k, _libssh2_ec_key *private_key,
 
 #endif /* LIBSSH2_ECDSA */
 
+#if LIBSSH2_ED25519
+
+int
+_libssh2_ed25519_sign(libssh2_ed25519_ctx *ctx, LIBSSH2_SESSION *session,
+                      uint8_t **out_sig, size_t *out_sig_len,
+                      const uint8_t *message, size_t message_len)
+{
+    int rc = -1;
+    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
+    size_t sig_len = 0;
+    unsigned char *sig = NULL;
+
+    if(md_ctx != NULL) {
+        if (EVP_DigestSignInit(md_ctx, NULL, NULL, NULL, ctx->private_key) != 1)
+            goto clean_exit;
+        if (EVP_DigestSign(md_ctx, NULL, &sig_len, message, message_len) != 1)
+            goto clean_exit;
+
+        if(sig_len != LIBSSH2_ED25519_SIG_LEN)
+            goto clean_exit;
+
+        sig = LIBSSH2_CALLOC(session, sig_len);
+        if(sig == NULL)
+            goto clean_exit;
+
+        rc = EVP_DigestSign(md_ctx, sig, &sig_len, message, message_len);
+    }
+
+    if(rc == 1) {
+        *out_sig = sig;
+        *out_sig_len = sig_len;
+    } else {
+        *out_sig_len = 0;
+        *out_sig = NULL;
+        LIBSSH2_FREE(session, sig);
+    }
+
+clean_exit:
+
+    if(md_ctx)
+        EVP_MD_CTX_free(md_ctx);
+
+    return (rc == 1 ? 0 : -1);
+}
+
+int
+_libssh2_curve25519_gen_k(_libssh2_bn **k,
+                          uint8_t private_key[LIBSSH2_ED25519_KEY_LEN],
+                          uint8_t server_public_key[LIBSSH2_ED25519_KEY_LEN])
+{
+    int rc = -1;
+    unsigned char out_shared_key[LIBSSH2_ED25519_KEY_LEN];
+    EVP_PKEY *peer_key = NULL, *server_key = NULL;
+    EVP_PKEY_CTX *server_key_ctx = NULL;
+    BN_CTX *bn_ctx = NULL;
+    size_t out_len = 0;
+
+    if(k == NULL)
+        return -1;
+
+    *k = NULL;
+
+    bn_ctx = BN_CTX_new();
+    if(bn_ctx == NULL)
+        return -1;
+
+    peer_key = EVP_PKEY_new_raw_public_key(EVP_PKEY_X25519, NULL,
+                                           server_public_key,
+                                           LIBSSH2_ED25519_KEY_LEN);
+
+    server_key = EVP_PKEY_new_raw_private_key(EVP_PKEY_X25519, NULL,
+                                              private_key,
+                                              LIBSSH2_ED25519_KEY_LEN);
+
+    if(peer_key == NULL || server_key == NULL) {
+        goto cleanExit;
+    }
+
+    server_key_ctx = EVP_PKEY_CTX_new(server_key, NULL);
+    if(server_key_ctx == NULL) {
+        goto cleanExit;
+    }
+
+    rc = EVP_PKEY_derive_init(server_key_ctx);
+    if(rc <= 0) goto cleanExit;
+
+    rc = EVP_PKEY_derive_set_peer(server_key_ctx, peer_key);
+    if(rc <= 0) goto cleanExit;
+
+    rc = EVP_PKEY_derive(server_key_ctx, NULL, &out_len);
+    if(rc <= 0) goto cleanExit;
+
+    if(out_len != LIBSSH2_ED25519_KEY_LEN) {
+        rc = -1;
+        goto cleanExit;
+    }
+
+    rc = EVP_PKEY_derive(server_key_ctx, out_shared_key, &out_len);
+
+    if(rc == 1 && out_len == LIBSSH2_ED25519_KEY_LEN) {
+        BN_bin2bn(out_shared_key, LIBSSH2_ED25519_KEY_LEN, *k);
```

Comment:
At this point, *k would be NULL (set in line 2634)

Maybe this line should be:
```C
        *k = BN_bin2bn(out_shared_key, LIBSSH2_ED25519_KEY_LEN, NULL);
```

Reply:
(or line 2634 should be removed if *k is already a BIGNUM)